renderdoc/driver/vulkan/vk_core.cpp

```diff
@@ -705,6 +705,59 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
   Threading::SetTLSValue(debugMessageSinkTLSSlot, (void *)sink);
 }
 
+void WrappedVulkan::AddPendingCommandBufferCallbacks(VkCommandBuffer commandBuffer)
+{
+  VkResourceRecord *cmdRecord = GetRecord(commandBuffer);
+
+  rdcarray<std::function<void()>> callbacks;
+  {
+    SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
+    if(cmdRecord->cmdInfo->pendingSubmissionCompleteCallbacks.callbacks.empty())
+      return;
+
+    callbacks.swap(cmdRecord->cmdInfo->pendingSubmissionCompleteCallbacks.callbacks);
+  }
+
+  const VkEventCreateInfo info = {VK_STRUCTURE_TYPE_EVENT_CREATE_INFO};
+  VkEvent event;
+  const VkResult vkr = ObjDisp(m_Device)->CreateEvent(Unwrap(m_Device), &info, NULL, &event);
+  CheckVkResult(vkr);
+
+  ObjDisp(commandBuffer)->CmdSetEvent(Unwrap(commandBuffer), event, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
+
+  VkPendingSubmissionCompleteCallbacks eventRef =
+      VkPendingSubmissionCompleteCallbacks(m_Device, event, std::move(callbacks));
+
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
+  cmdRecord->cmdInfo->pendingSubmissionCompleteCallbacks = eventRef;
+  m_PendingCmdBufferCallbacks.push_back(std::move(eventRef));
+}
+
+void WrappedVulkan::CheckPendingCommandBufferCallbacks()
+{
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
+
+  for(size_t i = 0; i < m_PendingCmdBufferCallbacks.size();)
+  {
+    VkPendingSubmissionCompleteCallbacks &pending = m_PendingCmdBufferCallbacks[i];
+
+    const VkResult vkr = ObjDisp(m_Device)->GetEventStatus(Unwrap(m_Device), pending);
+    if(vkr == VK_EVENT_SET)
+    {
+      for(std::function<void()> &f : pending.callbacks)
+        f();
+
+      m_PendingCmdBufferCallbacks.erase(i);
```

Comment:
If you're doing the loop like this (which is reasonable) you need to add a `continue` here to re-check `i` which will now be the next item in the list. As it is this will skip the element after each one erased.

Reply:
Yep, I used to have one and then accidently deleted it after some rearranging...